src/modm/ui/color/rgb.hpp

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright (c) 2009, Martin Rosekeit
+ * Copyright (c) 2009-2013, Fabian Greif
+ * Copyright (c) 2012-2013, 2015, Niklas Hauser
+ * Copyright (c) 2013, David Hebbeker
+ * Copyright (c) 2021, Thomas Sommer
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef MODM_COLOR_RGB_HPP
+#define MODM_COLOR_RGB_HPP
+
+#include <stdint.h>
+#include <bit>
+
+#include <concepts>
+#include <modm/io/iostream.hpp>
+#include <modm/math/utils/arithmetic_traits.hpp>
+#include <type_traits>
+
+#include "brightness.hpp"
+#include "hsv.hpp"
+#include "rgb565.hpp"
+
+#include <modm/debug/logger.hpp>
+
+namespace modm::color
+{
+
+// forward declarations for convertion constructors
+template<std::unsigned_integral T>
+class HsvT;
+
+template<std::unsigned_integral T>
+class BrightnessT;
+
+class Rgb565;
+
+/**
+ * Color in HSV Colorspace
+ *
+ * @author		Martin Rosekeit, Fabian Greif, Niklas Hauser, David Hebbeker, Thomas Sommer
+ * @ingroup		modm_ui_color
+ */
+template<std::unsigned_integral T>
+class RgbT
+{
+public:
+	T red{0};
+	T green{0};
+	T blue{0};
+
+	using TSum = modm::WideType<T>;
+
+	constexpr RgbT() = default;
+
+	constexpr RgbT(T red, T green, T blue) : red(red), green(green), blue(blue) {}
+
+	/**
+	 * Copy Constructor 8bit->16bit
+	 */
+	template<std::unsigned_integral U>
+	requires std::is_same_v<T, uint16_t> && std::is_same_v<U, uint8_t>
+    constexpr RgbT(const RgbT<U> &rgb_other)
+		: red(rgb_other.red << 8),
+		green(rgb_other.green << 8),
+		blue(rgb_other.blue << 8)
+	{}
+
+	/**
+	 * Copy Constructor 16bit->8bit
+	 */
+	template<std::unsigned_integral U>
+	requires std::is_same_v<T, uint8_t> && std::is_same_v<U, uint16_t>
+    constexpr RgbT(const RgbT<U> &rgb_other)
+		: red(rgb_other.red >> 8),
+		green(rgb_other.green >> 8),
+		blue(rgb_other.blue >> 8)
+	{}
```

Comment:
Added copy constructors  for 8bit -> 16bit and vice verse for all template color types. I've played around with alternative, more generic implementations but found this simple and readable ones using _concept overloading_ the best.

8 and 16 bit colors is feels like good enough for MCUs (until the day someone builds high-dynamic cameras with 24bits and more, haha)

Test added. If you like it ~~~> Please 📀 Merge